src/crafting.cpp

```diff
@@ -1746,219 +1791,277 @@ void set_item_inventory(item &newit)
     }
 }
 
+/* selection of component if a recipe requirement has multiple options (e.g. 'duct tap' or 'welder') */
+comp_selection player::select_component(const std::vector<item_comp> &components, int batch, bool can_cancel) {
+	std::vector<item_comp> player_has;
+	std::vector<item_comp> map_has;
+	std::vector<item_comp> mixed;
+
+	comp_selection selected = comp_selection();
+	inventory map_inv;
+	map_inv.form_from_map(pos3(), PICKUP_RANGE);
```

Comment:
For each list of components you create a new inventory containing all the items around your character. Have you tried this around a ***-load of items? That's why there is `player::crafting_inventory`. You may need to add an overload of that function for const player objects.